kong/runloop/handler.lua

```diff
@@ -646,14 +647,17 @@ return {
       end
 
       do
+        local balancer_data = ctx.balancer_data
+
+        -- set the upstream scheme as balancer selected
+        local upstream_scheme = balancer_data.scheme
+        var.upstream_scheme = upstream_scheme
```

Comment:
note 1: this is a breaking change for any plugin out there that relies on `ngx.var.upstream_scheme = "..."`, they now need to do `ngx.ctx.balancer_data.scheme = "..."`.

note 2: a good way to avoid such breaking changes would be to provide a PDK API for this operation (none currently exist in the `kong.service` module); this is outside the scope of this PR of course.

note 3: the `var.upstream_scheme = match_t.upstream_scheme` assignment in the `access.before` handler could arguably be removed, since plugins making use of it are already broken.

It is hard to know if such a change would have any impact - I myself am not sure if we offer any plugins making use of this variable (forward-proxy seems to hard-code `http://`, luckily). I do have a vague memory of it being mentioned from community users though.

Reply:
After some research, the PDK *does* have a function to modify the scheme. `request.set_scheme`.

Reply:
Indeed; after our irl this does mean that the patch in its current form breaks the PDK. Options were discussed, for now that is a blocker for this PR.